File: backend/src/presentation/api/routes/schedules.py
"""予定管理API - UseCase Pattern"""

from typing import Any

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from src.application.usecases.schedule_event_usecase import ScheduleEventUseCase
from src.presentation.api.dependencies import get_schedule_event_usecase

router = APIRouter(prefix="/schedules", tags=["schedules"])


class ScheduleEventCreateRequest(BaseModel):
    """予定作成リクエスト"""

    title: str
    date: str
    time: str
    type: str  # vaccination, outing, checkup, other
    location: str | None = None
    description: str | None = None
    status: str = "upcoming"  # upcoming, completed, cancelled
    created_by: str = "genie"
    user_id: str = "frontend_user"


class ScheduleEventUpdateRequest(BaseModel):
    """予定更新リクエスト"""

    title: str | None = None
    date: str | None = None
    time: str | None = None
    type: str | None = None
    location: str | None = None
    description: str | None = None
    status: str | None = None
    created_by: str | None = None
    user_id: str = "frontend_user"


@router.post("/create")
async def create_schedule_event(
    request: ScheduleEventCreateRequest,
    schedule_event_usecase: ScheduleEventUseCase = Depends(get_schedule_event_usecase),
) -> dict[str, Any]:
    """予定を作成"""
    try:
        request_data = request.model_dump()
        user_id = request_data.get("user_id", "frontend_user")

        result = await schedule_event_usecase.create_schedule_event(user_id=user_id, event_data=request_data)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/list")
async def get_schedule_events(
    user_id: str = "frontend_user",
    status: str | None = None,
    schedule_event_usecase: ScheduleEventUseCase = Depends(get_schedule_event_usecase),
) -> dict[str, Any]:
    """予定一覧を取得"""
    try:
        filters = {}
        if status:
            filters["status"] = status

        result = await schedule_event_usecase.get_schedule_events(user_id, filters)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/detail/{event_id}")
async def get_schedule_event(
    event_id: str,
    user_id: str = "frontend_user",
    schedule_event_usecase: ScheduleEventUseCase = Depends(get_schedule_event_usecase),
) -> dict[str, Any]:
    """特定の予定を取得"""
    try:
        result = await schedule_event_usecase.get_schedule_event(user_id, event_id)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.put("/update/{event_id}")
async def update_schedule_event(
    event_id: str,
    request: ScheduleEventUpdateRequest,
    schedule_event_usecase: ScheduleEventUseCase = Depends(get_schedule_event_usecase),
) -> dict[str, Any]:
    """予定を更新"""
    try:
        request_data = request.model_dump(exclude_unset=True)
        user_id = request_data.get("user_id", "frontend_user")

        result = await schedule_event_usecase.update_schedule_event(
            user_id=user_id, event_id=event_id, update_data=request_data,
        )
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/delete/{event_id}")
async def delete_schedule_event(
    event_id: str,
    user_id: str = "frontend_user",
    schedule_event_usecase: ScheduleEventUseCase = Depends(get_schedule_event_usecase),
) -> dict[str, Any]:
    """予定を削除"""
    try:
        result = await schedule_event_usecase.delete_schedule_event(user_id, event_id)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.patch("/status/{event_id}")
async def update_schedule_status(
    event_id: str,
    status: str,
    user_id: str = "frontend_user",
    schedule_event_usecase: ScheduleEventUseCase = Depends(get_schedule_event_usecase),
) -> dict[str, Any]:
    """予定のステータスを更新"""
    try:
        result = await schedule_event_usecase.update_schedule_status(user_id=user_id, event_id=event_id, status=status)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: frontend/src/libs/api/schedules.ts
/**
 * スケジュール管理API クライアント
 */

// APIベースURL
const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000'

// レスポンス型定義
interface ApiResponse<T = any> {
  success: boolean
  data?: T
  message?: string
}

// スケジュールイベントの型定義
export interface ScheduleEvent {
  id: string
  user_id: string
  title: string
  date: string
  time: string
  type: 'medical' | 'outing' | 'school' | 'other'
  location?: string
  description?: string
  status: 'upcoming' | 'completed' | 'cancelled'
  created_by: 'genie' | 'user'
  created_at: string
  updated_at: string
}

// 作成リクエストの型定義
export interface ScheduleEventCreateRequest {
  title: string
  date: string
  time: string
  type: 'medical' | 'outing' | 'school' | 'other'
  location?: string
  description?: string
  status?: 'upcoming' | 'completed' | 'cancelled'
  created_by?: 'genie' | 'user'
  user_id?: string
}

// 更新リクエストの型定義
export interface ScheduleEventUpdateRequest {
  title?: string
  date?: string
  time?: string
  type?: 'medical' | 'outing' | 'school' | 'other'
  location?: string
  description?: string
  status?: 'upcoming' | 'completed' | 'cancelled'
  created_by?: 'genie' | 'user'
  user_id?: string
}

/**
 * スケジュールイベント一覧を取得
 */
export async function getScheduleEvents(params?: {
  user_id?: string
  status?: string
}): Promise<ApiResponse<ScheduleEvent[]>> {
  try {
    const searchParams = new URLSearchParams()
    if (params?.user_id) searchParams.append('user_id', params.user_id)
    if (params?.status) searchParams.append('status', params.status)

    const url = `${API_BASE_URL}/api/schedules/list${searchParams.toString() ? `?${searchParams.toString()}` : ''}`

    const response = await fetch(url, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    })

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    const data = await response.json()
    return data
  } catch (error) {
    console.error('スケジュールイベント取得エラー:', error)
    return {
      success: false,
      message: error instanceof Error ? error.message : 'スケジュールイベントの取得に失敗しました',
    }
  }
}

/**
 * 特定のスケジュールイベントを取得
 */
export async function getScheduleEvent(
  eventId: string,
  userId: string = 'frontend_user'
): Promise<ApiResponse<ScheduleEvent>> {
  try {
    const response = await fetch(
      `${API_BASE_URL}/api/schedules/detail/${eventId}?user_id=${userId}`,
      {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        },
      }
    )

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    const data = await response.json()
    return data
  } catch (error) {
    console.error('スケジュールイベント詳細取得エラー:', error)
    return {
      success: false,
      message: error instanceof Error ? error.message : 'スケジュールイベントの取得に失敗しました',
    }
  }
}

/**
 * 新しいスケジュールイベントを作成
 */
export async function createScheduleEvent(
  eventData: ScheduleEventCreateRequest
): Promise<ApiResponse<ScheduleEvent>> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/schedules/create`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        ...eventData,
        user_id: eventData.user_id || 'frontend_user',
      }),
    })

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    const data = await response.json()
    return data
  } catch (error) {
    console.error('スケジュールイベント作成エラー:', error)
    return {
      success: false,
      message: error instanceof Error ? error.message : 'スケジュールイベントの作成に失敗しました',
    }
  }
}

/**
 * スケジュールイベントを更新
 */
export async function updateScheduleEvent(
  eventId: string,
  eventData: ScheduleEventUpdateRequest
): Promise<ApiResponse<ScheduleEvent>> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/schedules/update/${eventId}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        ...eventData,
        user_id: eventData.user_id || 'frontend_user',
      }),
    })

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    const data = await response.json()
    return data
  } catch (error) {
    console.error('スケジュールイベント更新エラー:', error)
    return {
      success: false,
      message: error instanceof Error ? error.message : 'スケジュールイベントの更新に失敗しました',
    }
  }
}

/**
 * スケジュールイベントを削除
 */
export async function deleteScheduleEvent(
  eventId: string,
  userId: string = 'frontend_user'
): Promise<ApiResponse> {
  try {
    const response = await fetch(
      `${API_BASE_URL}/api/schedules/delete/${eventId}?user_id=${userId}`,
      {
        method: 'DELETE',
        headers: {
          'Content-Type': 'application/json',
        },
      }
    )

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    const data = await response.json()
    return data
  } catch (error) {
    console.error('スケジュールイベント削除エラー:', error)
    return {
      success: false,
      message: error instanceof Error ? error.message : 'スケジュールイベントの削除に失敗しました',
    }
  }
}

/**
 * スケジュールイベントのステータスを更新
 */
export async function updateScheduleEventStatus(
  eventId: string,
  status: 'upcoming' | 'completed' | 'cancelled',
  userId: string = 'frontend_user'
): Promise<ApiResponse<ScheduleEvent>> {
  try {
    const response = await fetch(
      `${API_BASE_URL}/api/schedules/status/${eventId}?user_id=${userId}`,
      {
        method: 'PATCH',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ status }),
      }
    )

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    const data = await response.json()
    return data
  } catch (error) {
    console.error('スケジュールイベントステータス更新エラー:', error)
    return {
      success: false,
      message: error instanceof Error ? error.message : 'ステータスの更新に失敗しました',
    }
  }
}
